### PR TITLE
Refactor dataloaders

### PR DESCRIPTION
**Pull request by TibbersHao**
This PR modified the place where qlty stitching should happen during training and inference.

To specify:
1. During training: all labeled frames and their corresponding masks will be loaded into memory and cropped into patches by `qlty`, right now the `batch_size_train` controls how many patches we want to load into device per batch. Previously this parameter controls the number of frames loaded per batch, which will likely run into memory issues with large patch numbers for big images.

2. During inference: grab a single frame at a time, crop to patches, then use `batch_size_inference` to control how many patches we want to make prediction per batch. Previously this parameter was set to control how many images we pass into device per batch, which will give same issue as described above.

3. Added data standardization for input images.

### CODE REVIEW

File: src/seg_utils.py [hunk 1]
```diff
@@ -363,4 +420,34 @@ def segment(net, device, inference_loader, qlty_object, tiled_client):
...
                 frame_number+=1
 
-    return frame_number
\ No newline at end of file
+    return frame_number
+
+def crop_seg_save(net, device, image, qlty_object, parameters, tiled_client, frame_idx):    
+    assert image.ndim==2
+    # Standardization
+    image = (image - np.mean(image)) / np.std(image)
```
Comment (taxe10) [inline src/seg_utils.py:428]: When applying standardization, it is necessary to use the same mean and variance for both all sets (training, validation, and testing). My suggestion would be to add the mean and variance to the model metadata so it can be used for inference later.

**Review by taxe10 — CHANGES_REQUESTED** [review]

File: src/seg_utils.py (hunk 1 above)
Comment (TibbersHao) [reply src/seg_utils.py:428]: Sounds good. I thought the reason of doing standardization is to bring all data set to be confined with a fixed mean and std. But your point does make sense. Working on that right now.

Comment (TibbersHao) [reply src/seg_utils.py]: Update and wrap this comment up: right now we are doing pixel / 255 across frames, and I've added a TODO in code for future discussion of normalization.

File: src/seg_utils.py [hunk 2]
```diff
@@ -63,6 +65,66 @@ def train_val_split(dataset, parameters):
...
+    )  # TODO: Revisit the best normalization option we can do, thinking about transferring between sets and train-inference.
+    images = torch.from_numpy(images)
+    masks = torch.from_numpy(masks)
+
+    qlty_window = parameters.qlty_window
+    qlty_step = parameters.qlty_step
+    qlty_border = parameters.qlty_border
+
+    if images.ndim == 3:
+        images = images.unsqueeze(1)
```
Comment (Wiebke) [inline src/seg_utils.py:81]: This looks a bit suspicious. Could you explain when this case occurs and why we are added a dimension at `axis=1` rather than 0?

File: src/train.py [hunk 3]
```diff
@@ -61,12 +62,19 @@
         mask_tiled_uri=io_parameters.mask_tiled_uri,
         mask_tiled_api_key=io_parameters.mask_tiled_api_key,
         is_training=True,
+        using_qlty=False,
```
Comment (Wiebke) [inline src/train.py:65]: Are we introducing this variable to retain the possibility to go back to the old version of the TiledDataSet, it seems to always be `using_qlty=False`?

File: src/utils.py [hunk 4]
```diff
@@ -56,26 +58,32 @@ def allocate_array_space(
...
 
     last_container = last_container.create_container(key=uid)
+
     array_shape = (
         tiled_dataset.mask_client.shape
         if tiled_dataset.mask_client
         else tiled_dataset.data_client.shape
     )
+    # In case we have 2d array for the single mask case, where the ArrayClient will return a 2d array.
+    if len(array_shape) == 2:
```
Comment (Wiebke) [inline src/utils.py:73]: Did you encounter this case when writing a single mask frame through the dash interface? If yes, I should investigate that code again.

File: src/utils.py [hunk 5]
```diff
@@ -34,7 +34,9 @@ def ensure_parent_containers(tiled_uri, tiled_api_key):
 
     last_container = from_uri(tiled_root, api_key=tiled_api_key)
 
-    container_parts = path_parts[3:]
+    container_parts = path_parts[
```
Comment (Wiebke) [inline src/utils.py:37]: Assume we have not specified a subpath but are just writing to the base uri http://localhost:8000/, does the function here still do what we want it to do? 
It seems we would run into errors since this assumes that path_parts will have at least 4 elements?

**Review by Wiebke — COMMENTED** [review]
Thank you all for your work on this. 
We successfully segmented data today with this refactor.

I do have a few questions though, where I don't quite follow the code.

File: src/utils.py (hunk 4 above)
Comment (taxe10) [reply src/utils.py:73]: No, we manually ingested the annotations in tiled for testing yesterday. I accidentally wrote the mask as a 2D array since it was just 1 frame. If the dash interface saves the masks as a 3D array even when there is a single frame, we can remove these lines of code.

File: src/seg_utils.py (hunk 2 above)
Comment (TibbersHao) [reply src/seg_utils.py:81]: Sure. This statement is likely to be triggered from the current pipeline. The previous step forges all labeled frames into a numpy nd array. Since each slice we pulled from tiled data_client is in shape of eg: (2560, 2560), this array will be 3d, eg: (3, 2560, 2560). qlty will require the input array to be converted to a tensor of shape (N, C, Y, X), so we need to add the channel information and make that conversion, that's why we are doing `axis=1` instead of 0.

In an earlier commit I did use consecutive `unsqueeze(0)`, and ended with a bug of tensor shape (1, 3, 2560, 2560).

File: src/train.py (hunk 3 above)
Comment (TibbersHao) [reply src/train.py:65]: Yes exactly. This optional parameter is hardcoded to turn off all qlty related operation in `TiledDataset`, while not losing its old functionality, in case we change our mind in the future.

File: src/utils.py (hunk 4 above)
Comment (TibbersHao) [reply src/utils.py:73]: Yes, in one of our testing case yesterday, @taxe10 manually ingested some data into the local Tiled Server, and there was only 1 mask for the `mask_client`. We ended up with `dataset.mask_client` resulted in a shape of 2d array structure, and ran into "index of out bound" issues when retrieving the block information. So this is a layer added to mitigate the issue.

Please keep an eye to see whether this is also the case for the current pipeline, if we ingest a single mask from the dash app.

File: src/utils.py (hunk 5 above)
Comment (TibbersHao) [reply src/utils.py:37]: You are right. With the case you suggested above, this will fail. I recall from last Friday's discussion, the `api/v1/metadata` will always be the pre-existing part once a tiled client is installed and ready for the ingestion. And I believe we made a decision of going forward with this assumption due to the limited time.

We can revisit this again later, but having `[3:]` here will prevent the duplication of the above path for the current MVP.

File: src/seg_utils.py (hunk 2 above)
Comment (Wiebke) [reply src/seg_utils.py:81]: I see, makes sense.

File: src/utils.py (hunk 4 above)
Comment (Wiebke) [reply src/utils.py:73]: Makes sense. This is not the result when we are exporting the mask from the interface.
